plugin/trino-hive/src/main/java/io/trino/plugin/hive/util/TextHeaderWriter.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.util;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.plugin.hive.HiveType;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeManager;
+import org.apache.hadoop.hive.serde2.SerDeException;
+import org.apache.hadoop.hive.serde2.Serializer;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspectorFactory;
+import org.apache.hadoop.hive.serde2.objectinspector.StandardStructObjectInspector;
+import org.apache.hadoop.io.BinaryComparable;
+import org.apache.hadoop.io.Text;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.hive.HiveSessionProperties.getTimestampPrecision;
+
+public class TextHeaderWriter
+{
+    private final Serializer serializer;
+    private final Type headerType;
+    private final List<String> fileColumnNames;
+
+    public TextHeaderWriter(Serializer serializer, TypeManager typeManager, ConnectorSession session, List<String> fileColumnNames)
+    {
+        this.serializer = serializer;
+        this.fileColumnNames = fileColumnNames;
+        this.headerType = HiveType.valueOf("string").getType(typeManager, getTimestampPrecision(session));
+    }
+
+    public void write(final OutputStream compressedOutput, int rowSeparator)
+            throws IOException
+    {
+        try {
+            ObjectInspector stringObjectInspector = HiveWriteUtils.getRowColumnInspector(headerType);
+            ImmutableList<Text> headers = fileColumnNames.stream().map(Text::new).collect(toImmutableList());
+            List<ObjectInspector> inspectors = IntStream.range(0, fileColumnNames.size()).mapToObj(ignored -> stringObjectInspector).collect(toImmutableList());
+            StandardStructObjectInspector headerStructObjectInspectors = ObjectInspectorFactory.getStandardStructObjectInspector(fileColumnNames, inspectors);
+            BinaryComparable binary = (BinaryComparable) serializer.serialize(headers, headerStructObjectInspectors);
```

Comment:
This may be ok, but feels incorrect.
I am not an expert in Hive SerDe code but here we are constructing `serializer` passing actual table column types as parameters:
```java
        String serDe = storageFormat.getSerDe();
        serializer = initializeSerializer(conf, schema, serDe);
```

And then we are using it for writing header which consists of just strings.

Maybe we can make it simpler. It seems to me that we may skip introducing  `TextHeaderWriter` at all.
In `HiveWriteUtils.java` where we are constructing `TextRecordWriter` we have all needed information already.
We can add an extra `Optional<List<String>> headerColumnNames` parameter to `TextRecordWriter` constructor.
And in `HiveWriteUtils.createRecordWriter` we can do sth like:

```java
          if (outputFormatName.equals(HiveIgnoreKeyTextOutputFormat.class.getName())) {
                List<String> fileColumnNames = getColumnNames(properties);
                return new TextRecordWriter(target, conf, properties, compress, fileColumnNames);
            }
```

Then we directly write a single extra line to `outputStream` in `TextRecordWriter`, not using any hive voodoo for that. Just create a byte buffer from given list of column names, and write it down.

Wouldn't that work?

Would it be worse?

Reply:
Regarding the incorrectness that may come from the way serializer is initialized I checked that in serialize method only objectInspector that is passed is used and this one is correct as I initialize it myself.

I'd love to make it simpler! I considered using exactly the same approach you suggest but I didn't go that way for two reasons:
 1.  I assumed that there can be configured specific "column separator" character that should be used between different column names - I didn't find that information in available properties 
 2. to correctly convert string into byte buffer I would need to know what charset should be used - again I didn't find that information in available properties

Apart from those 2 reasons I also  thought that it is cleaner to use the same mechanism to write this header instead of doing it on the side.

Please let me know if my concerns are not valid or if you see any obvious way to satisfy them.

Reply:
> Regarding the incorrectness that may come from the way serializer is initialized I checked that in serialize method only objectInspector that is passed is used and this one is correct as I initialize it myself.

Indeed. It look like object inspectors stored within serde are only used on read flow.

> 
> I'd love to make it simpler! I considered using exactly the same approach you suggest but I didn't go that way for two reasons:
> 
> 1. I assumed that there can be configured specific "column separator" character that should be used between different column names - I didn't find that information in available properties

Make sense. Yet - should the column separator for header and data be the same. Possibly - but I am not super super sure - not that we care much :)

> 2. to correctly convert string into byte buffer I would need to know what charset should be used - again I didn't find that information in available properties

Fair. Though I am not sure if there can be any non-ascii-letter chars in column names.

> 
> Apart from those 2 reasons I also thought that it is cleaner to use the same mechanism to write this header instead of doing it on the side.
> 
> Please let me know if my concerns are not valid or if you see any obvious way to satisfy them.

Let's keep it then.

Reply:
ok